### ISSUE TITLE
Async url provider

### ISSUE THREAD

**Issue by padge** [robtaussig/react-use-websocket#33]
It would be cool if the first `url` param could optionally be an async function that returns a URL, similar to what's found in this package https://www.npmjs.com/package/reconnecting-websocket#update-url. 

My use case is similar to the example they give, fetching a short-lived auth token:
```js
const urlProvider = async () => {
    const token = await getSessionToken();
    return `wss://my.site.com/${token}`;
};
```

**Comment by codeocelot** [comment]
I have the same pain-point, however, I can't really see a good way to add this to the project.  As a work around, I've placed `getSessionToken()` in a parent component and only render the component when a jwt is acquired.  Not ideal, but works pretty well all things given.  Best of luck!

**Comment by robtaussig** [comment]
Great question @padge, and great suggestion for a work-around @codeocelot.

Since there is no guarantee that the websocket is immediately available to send messages (as you still need to wait for readyState to be open), there should not be any expectation of synchronous access to sendMessage or getWebSocket (and for the latter, we can return null or something until it's ready). Therefore, I don't suspect too much issue with this. That said, it wouldn't be the first time I underestimated the complexity of a solution, so I'll have to think a bit more and get back on this.

**Comment by padge** [comment]
@codeocelot thanks for the workaround suggestion! Unfortunately I wasn't able to make it work nicely when reconnecting, since I use a short-lived JWT that needs to be re-fetched on reconnect.

**Comment by codeocelot** [comment]
@padge are you sending your jwt as a query param?  You can try appending it to the url yourself instead of using the query option (which needs to be static).  If you do this, each time `useWebsocket` gets a new url it'll handle switching to the new connection.

As a side note, I think jwt refreshes are a good example of when options shouldn't be considered static.

**Comment by robtaussig** [comment]
@codeocelot I agree that this is one of several examples of why options should not be considered static. I'm testing a build out now where options is treated as a dynamic property and hopefully can have it ready soon.

**Comment by robtaussig** [comment]
Hi @padge,

I merged in the change that supports passing an async url getter to useWebSocket in lieu of a string, updated the documentation, and bumped the library to 1.4.0.

bfe7f14

Thank you for your feedback and patience!

**Comment by padge** [comment]
Right on - thanks!

**Comment by padge** [comment]
Figured I'd just comment here for now rather than create a new issue.

Can the url provider be called on reconnect? I'm using a short-lived token and reconnecting currently uses the old expired auth token.

**Comment by robtaussig** [comment]
@padge,

Absolutely -- that was an oversight that I didn't do that in the first place.

**Comment by robtaussig** [comment]
@padge,

Thanks for bringing this to my attention. Pushed out the fix and bumped version to 1.6.2.

caa3f0
